lib/puppet/face/facts.rb

```diff
@@ -35,4 +35,43 @@
 
   deactivate_action(:destroy)
   deactivate_action(:search)
+
+  action(:upload) do
+    summary _("Upload local facts to the puppet master.")
+    description <<-'EOT'
+      Reads facts from the local system using the `facter` terminus, then
+      saves the returned facts using the rest terminus.
+    EOT
+    returns "Nothing."
+    notes <<-'EOT'
+      This action requires that the puppet master's `auth.conf` file
+      allow save access to the `facts` REST terminus. Puppet agent does
+      not use this facility, and it is turned off by default. See
+      <https://puppet.com/docs/puppetserver/latest/config_file_auth.html>
+      for more details.
+    EOT
+    examples <<-'EOT'
+      Upload facts:
+
+      $ puppet facts upload
+    EOT
+
+    render_as :json
+
+    when_invoked do |options|
+      # Use `agent` sections  settings for certificates, Puppet Server URL,
+      # etc. instead of `user` section settings.
+      Puppet.settings.preferred_run_mode = :agent
+      Puppet::Node::Facts.indirection.terminus_class = :facter
+      facts = Puppet::Node::Facts.indirection.find(Puppet[:certname])
```

Comment:
I think we want to use `Puppet[:node_name_value]` which defaults to `$certname` but it doesn't have to be the same. Also `node_name_value` is what the agent uses when requesting node & catalog and sending reports.

Reply:
Hmm, seems like a good idea --- but I can't recall off the top of my head whether `node_name_value` is dependent on `node_name_fact`, which requires loading facts, which puts us in a bit of a Catch-22. Also, the node name used for the lookup might not even matter when the terminus is set to `:facter`.

Definitely seems like the thing to use when POSTing facts back to the server though. Maybe just need to ensure the `name` attribute of the returned factset is populated before the save call.

Reply:
`node_name_value` and `node_name_fact` are mutually exclusive. Using the fact-based one definitely has problems. You're right that the name we request does not matter when using the `Facter` terminus, except that it becomes the `name` attribute of the returned facts object. So I think we want to use `node_name_value` so that the returned facts object is consistent with what puppet thinks it's name is.

Reply:
Looks like the common pattern used by `apply` and agent is to query using `Puppet[:node_name_value]` and then re-set `Puppet[:node_name_value]` and the `name` attribute of the returned factset if `Puppet[:node_name_fact]` is not empty:

https://github.com/puppetlabs/puppet/blob/5.3.4/lib/puppet/application/apply.rb#L204-L212
https://github.com/puppetlabs/puppet/blob/5.3.4/lib/puppet/configurer/fact_handler.rb#L15-L19

I'll implement the strategy from `fact_handler.rb` and add some test coverage.